cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -293,6 +303,90 @@ Status DiscoverFilesFromDir(const std::shared_ptr<fs::LocalFileSystem>& local_fs
   return Status::OK();
 }
 
+namespace internal {
```

Comment:
I would be leaning towards removing the internal namespace here. I don't see this being different from other methods here. Plus the file is already named "relation_internal" so methods here either all belongs to an internal namespace, or don't, but I don't see a reason to split them.

Reply:
My [rationale is explained here](https://github.com/apache/arrow/pull/34627#discussion_r1142728086). I'm fine with removing the namespace, if you feel strongly about it.

Reply:
Yeah I prefer that we remove this namespace.

Reply:
Actually I see what you are doing here - these two methods are not "API" methods in the same sense that the other "FromProto" methods are and takes a lot of internal parameters. 

We can leave as is. But can you put a comment about how the methods here are different from the rest of the methods here. I didn't understand why do meant by "Called sparsely" - but I guess you mean that "these two methods are only meant to be used inside FromProto of AggregationRel and SegementedAggregationRel"?

Reply:
What is confusing here is that a "internal" in the filename and "internal" in the namespace are two difference things. Here the namespace "internal" means "internals of this internal cc file", which is not obvious. Perhaps we should name this to engine::aggregate to make clear this is internal to aggregate logic.

Reply:
Actually I think if we remove these two methods from the header I think there is no need to make this internal namespace either.